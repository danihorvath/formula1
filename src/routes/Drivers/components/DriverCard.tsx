import Card from "@mui/material/Card";
import Box from "@mui/material/Box";
import Typography from "@mui/material/Typography";
import CardMedia from "@mui/material/CardMedia";
import CardHeader from "@mui/material/CardHeader";
import CardActions from "@mui/material/CardActions";
import IconButton from "@mui/material/IconButton";
import Avatar from "@mui/material/Avatar";
import { Driver } from "../index";
import ArrowUpwardIcon from "@mui/icons-material/ArrowUpward";

interface DriverCardProps {
  driver: Driver;
  overtake: (id: number) => void;
}

const DriverCard = ({ driver, overtake }: DriverCardProps) => (
  <Card variant="outlined" key={driver.id}>
    <CardHeader
      avatar={<Avatar aria-label="recipe">{driver.place}</Avatar>}
      title={`${driver.firstname} ${driver.lastname}`}
      subheader={driver.team}
      action={
        <Box display="flex" height={50} alignItems="center" marginX={1}>
          <Typography fontWeight={500}>{driver.code}</Typography>
        </Box>
      }
    />
    <CardMedia
      component="img"
      image={`${import.meta.env.VITE_API_URL}${driver.imgUrl}`}
      alt="Paella dish"
    />
    <CardActions>
      <IconButton
        aria-label="add to favorites"
        sx={{ marginLeft: "auto" }}
        disabled={driver.place === 1}
        onClick={() => overtake(driver.id)}
      >
        <ArrowUpwardIcon />
      </IconButton>
    </CardActions>
  </Card>
);

export default DriverCard;
