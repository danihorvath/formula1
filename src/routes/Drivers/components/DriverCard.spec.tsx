import { expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import * as matchers from "@testing-library/jest-dom/matchers";
import "@testing-library/jest-dom/vitest";
import DriverCard from "./DriverCard";

expect.extend(matchers);

const mock = {
  driver: {
    id: 23,
    code: "HAM",
    firstname: "Lewis",
    lastname: "Hamilton",
    country: "United Kingdom",
    team: "Mercedes",
    place: 2,
    imgUrl: "/images/hamilton.jpg",
  },
  overtake: vi.fn(),
};

test("Snapshot Test", () => {
  const { container } = render(
    <DriverCard driver={mock.driver} overtake={mock.overtake} />
  );
  expect(container).toMatchSnapshot();
});

test("Displays the name of the driver", () => {
  render(<DriverCard driver={mock.driver} overtake={mock.overtake} />);
  expect(screen.getByTestId("header")).toHaveTextContent(
    `${mock.driver.firstname} ${mock.driver.lastname}`
  );
});

test("Displays the team", () => {
  render(<DriverCard driver={mock.driver} overtake={mock.overtake} />);
  expect(screen.getByTestId("header")).toHaveTextContent(mock.driver.team);
});

test("Displays the current place", () => {
  render(<DriverCard driver={mock.driver} overtake={mock.overtake} />);
  expect(screen.getByTestId("header")).toHaveTextContent(
    mock.driver.place.toString()
  );
});

test("Displays the code", () => {
  render(<DriverCard driver={mock.driver} overtake={mock.overtake} />);
  expect(screen.getByTestId("header")).toHaveTextContent(mock.driver.code);
});

test("Displays the photo", () => {
  render(<DriverCard driver={mock.driver} overtake={mock.overtake} />);
  expect(screen.getByRole("img")).toHaveAttribute(
    "src",
    `${import.meta.env.VITE_API_URL}${mock.driver.imgUrl}`
  );
});

test("Overtake method not called for first driver", () => {
  const spy = vi.spyOn(mock, "overtake");
  render(
    <DriverCard
      driver={{ ...mock.driver, place: 1 }}
      overtake={mock.overtake}
    />
  );
  screen.getByRole("button").click();
  expect(spy).toHaveBeenCalledTimes(0);
});

test("Calls overtake method with the driver id", () => {
  const spy = vi.spyOn(mock, "overtake");
  render(<DriverCard driver={mock.driver} overtake={mock.overtake} />);

  screen.getByTestId("button").click();

  expect(spy).toHaveBeenCalledWith(mock.driver.id);
});
