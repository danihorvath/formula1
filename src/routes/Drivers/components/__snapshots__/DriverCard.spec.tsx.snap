// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot Test 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
      data-testid="header"
    >
      <div
        class="MuiCardHeader-avatar css-1ssile9-MuiCardHeader-avatar"
      >
        <div
          aria-label="recipe"
          class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-10m8oxi-MuiAvatar-root"
        >
          2
        </div>
      </div>
      <div
        class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiTypography-body2 MuiCardHeader-title css-et1ao3-MuiTypography-root"
        >
          Lewis Hamilton
        </span>
        <span
          class="MuiTypography-root MuiTypography-body2 MuiCardHeader-subheader css-83ijpv-MuiTypography-root"
        >
          Mercedes
        </span>
      </div>
      <div
        class="MuiCardHeader-action css-sgoict-MuiCardHeader-action"
      >
        <div
          class="MuiBox-root css-ctup6s"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-eflgtu-MuiTypography-root"
          >
            HAM
          </p>
        </div>
      </div>
    </div>
    <img
      alt="Paella dish"
      class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
      src="http://localhost:3000/images/hamilton.jpg"
    />
    <div
      class="MuiCardActions-root MuiCardActions-spacing css-dnrpxu-MuiCardActions-root"
    >
      <button
        aria-label="add to favorites"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-djcg5u-MuiButtonBase-root-MuiIconButton-root"
        data-testid="button"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ArrowUpwardIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="m4 12 1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
